lib/Transforms/VectorDistribution.cpp

```diff
@@ -20,6 +20,205 @@ using namespace mlir;
 using namespace mlir::linalg;
 
 namespace {
+
+/// Modified version of the upstream pattern handling more cases.
+/// TODO: Replace upstream once it is ready.
+struct TransferReadExtractPattern
+    : public OpRewritePattern<vector::ExtractMapOp> {
+  TransferReadExtractPattern(MLIRContext *context)
+      : OpRewritePattern<vector::ExtractMapOp>(context) {}
+  LogicalResult matchAndRewrite(vector::ExtractMapOp extract,
+                                PatternRewriter &rewriter) const override {
+    auto read = extract.vector().getDefiningOp<vector::TransferReadOp>();
+    if (!read || read.mask()) return failure();
+
+    SmallVector<Value, 4> indices(read.indices().begin(), read.indices().end());
+    AffineMap indexMap = extract.map().compose(read.permutation_map());
+    unsigned idCount = 0;
+    // We cannot pull down transfer_read ops as it may cross ops with
+    // side-effects.
+    rewriter.setInsertionPoint(read);
+    ImplicitLocOpBuilder lb(read.getLoc(), rewriter);
```

Comment:
no builder within a rewriter please, messes up the tracking (and our future drivers) IIRC

Reply:
removed